extensions/amp-viewer-integration/0.1/messaging.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {listen} from '../../../src/event-helper';
+import {dev} from '../../../src/log';
+
+const TAG = 'amp-viewer-messaging';
+const sentinel_ = '__AMPHTML__';
+const requestSentinel_ = sentinel_ + 'REQUEST';
+const responseSentinel_ = sentinel_ + 'RESPONSE';
+
+/**
+ * @fileoverview This is used in amp-viewer-integration.js for the
+ * communication protocol between AMP and the viewer. In the comments, I will
+ * refer to the communication as a conversation between me and Bob. The
+ * messaging protocol should support both sides, but at this point I'm the
+ * ampdoc and Bob is the viewer.
+ */
+export class Messaging {
+
+  /**
+   * Conversation (messaging protocol) between me and Bob.
+   * @param {!Window} source
+   * @param {!Window} target
+   * @param {string} targetOrigin
+   * @param {function(string, *, boolean):(!Promise<*>|undefined)}
+   *    requestProcessor
+   */
+  constructor(source, target, targetOrigin, requestProcessor) {
+    /**  @private {!number} */
+    this.requestIdCounter_ = 0;
+    /**  @private {!Object<string, {resolve: function(*), reject: function(!Error)}>} */
+    this.waitingForResponse_ = {};
+    /** @const @private {!Window} */
+    this.target_ = target;
+    /** @const @private {string} */
+    this.targetOrigin_ = targetOrigin;
+    /** @const @private {function(string, *, boolean):(!Promise<*>|undefined)} */
+    this.requestProcessor_ = requestProcessor;
+
+    dev().assert(this.targetOrigin_, 'Target origin must be specified!');
+
+    listen(source, 'message', this.handleMessage_.bind(this));
+  }
+
+  /**
+   * Bob sent me a message. I need to decide if it's a new request or
+   * a response to a previous 'conversation' we were having.
+   * @param {?Event} event
+   * @private
+   */
+  handleMessage_(event) {
+    if (!event || event.source != this.target_ ||
```

Comment:
`!event` check can now be removed and `?Event` can be changed to `!Event` above.